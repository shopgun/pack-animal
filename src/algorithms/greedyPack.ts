import { IPoint, IVerifyPackOptions, rotateMatrixAroundPoint, verifyPack } from "../geometry";
import { getPolygonTransform, ITransform } from "../transform";
import { Matrix } from "../vendor/matrix";

export enum RotationMode {
  Off,
  Simple = "SIMPLE",
  Advanced = "ADVANCED"
}

export interface IGreedyPackOptions {
  rotationMode?: RotationMode;
  polygonHitboxScale?: number;
  rectangleHitboxScale?: number;
}

enum MatrixAttribute {
  translateX = "translateX",
  translateY = "translateY"
}
const maxMatrix = (
  attribute: MatrixAttribute,
  matrix: Matrix,
  increment: number,
  points: IPoint[],
  verifier: (transformedPoints: IPoint[]) => boolean
): number => {
  let i = 0;
  let previousValue = null;
  let transformedPoints: IPoint[];
  const { scale, translate: { x, y } } = matrix.decompose();
  const initial = attribute === MatrixAttribute.translateX ? x : y;
  let value = initial;
  do {
    previousValue = value;
    value = initial + increment * i;
    const m = new Matrix()
      .translate(x, y)
      [attribute](value)
      .scaleU(scale.x);
    transformedPoints = m.applyToArray(points);
    i++;
  } while (verifier(transformedPoints));
  return previousValue;
};

export const greedyPack = (
  rectangleWidth: number,
  rectangleHeight: number,
  polygons: IPoint[][],
  { rotationMode = RotationMode.Simple, polygonHitboxScale }: IGreedyPackOptions = {}
): ITransform[] => {
  if (!polygons.length) {
    return [];
  }
  const rectangle: IPoint[] = [{ x: 0, y: 0 }, { x: rectangleWidth, y: rectangleHeight }];
  let polygonTransforms;
  const scaleIncrement = 0.025;
  const translateXIncrement = rectangleWidth * 0.01;
  const translateYIncrement = rectangleHeight * 0.01;
  const rotateIncrement = 1;

  const scaleInitial = 1;
  const rotateInitial = 0;
  const translateXInitial = 0;
  const translateYInitial = 0;
  let scale: number;
  let j = 0;
  const verifyPackOptions: IVerifyPackOptions = {
    polygonHitboxScale
  };
  do {
    scale = scaleInitial - scaleIncrement * j;
    polygonTransforms = polygons.reduce((memo: ITransform[], points: IPoint[]): ITransform[] => {
      const memoPoints = memo.map(transformPoly => transformPoly.points);
      const verifier = (newPoints: IPoint[]) =>
        verifyPack([...memoPoints, newPoints], rectangle, verifyPackOptions);

      let rotate = rotateInitial;
      let translateX = translateXInitial;
      let translateY = translateYInitial;
      let transformedPoints;
      let i;
      translateY = maxMatrix(
        MatrixAttribute.translateY,
        new Matrix().translate(translateX, translateY).scaleU(scale),
        translateYIncrement,
        points,
        verifier
      );

      translateX = maxMatrix(
        MatrixAttribute.translateX,
        new Matrix().translate(translateX, translateY).scaleU(scale),
        translateXIncrement,
        points,
        verifier
      );

      let finalMatrix = new Matrix();
      if (rotationMode === RotationMode.Advanced) {
        finalMatrix = new Matrix()
          .translate(translateX, translateY)
          .scale(scale, scale)
          .applyToArray(points)
          .reduce((memoMatrix, point) => {
            let previousRotate = null;
            i = 0;
            do {
              previousRotate = rotate;
              rotate = rotateInitial + rotateIncrement * i;
              const m = rotateMatrixAroundPoint(point, rotate, memoMatrix);
              m.multiply(new Matrix().translate(translateX, translateY).scale(scale, scale));
              transformedPoints = m.applyToArray(points);
              i++;
            } while (verifier(transformedPoints) && rotate <= 22.5);

            return rotateMatrixAroundPoint(point, previousRotate, memoMatrix);
          }, new Matrix());
      }

      if (rotationMode === RotationMode.Simple) {
        const width = Math.floor(
          Math.max(...points.map(point => point.x)) - Math.min(...points.map(point => point.x))
        );
        const height = Math.floor(
          Math.max(...points.map(point => point.y)) - Math.min(...points.map(point => point.y))
        );
        const center = { x: width / 2, y: height / 2 };
        let previousRotate = null;
        i = 0;
        do {
          previousRotate = rotate;
          rotate = rotateInitial + rotateIncrement * i;
          const m = rotateMatrixAroundPoint(center, rotate);
          m.multiply(new Matrix().translate(translateX, translateY).scale(scale, scale));
          transformedPoints = m.applyToArray(points);
          i++;
        } while (verifier(transformedPoints));

        finalMatrix = rotateMatrixAroundPoint(center, previousRotate);
      }

      finalMatrix.translate(translateX, translateY).scale(scale, scale);

      memo.push(getPolygonTransform(rectangleWidth, rectangleHeight, points, finalMatrix));
      return memo;
    }, []);
    j++;
  } while (
    !verifyPack(polygonTransforms.map(({ points }) => points), rectangle, verifyPackOptions)
  );

  return polygonTransforms;
};
