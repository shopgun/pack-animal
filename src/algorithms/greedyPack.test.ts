import { RotationMode } from "./";
import { greedyPack } from "./greedyPack";

describe("greedyPack", () => {
  it("works if true is truthy", () => {
    expect(true).toBeTruthy();
  });
  it("gives nothing for nothingg", () => {
    const polygonTransforms = greedyPack(0, 0, []);
    expect(polygonTransforms).toMatchSnapshot();
  });
  it("works", () => {
    const polygonTransforms = greedyPack(
      800,
      742,
      [
        [{ x: 1, y: 1 }, { x: 1, y: 366 }, { x: 798, y: 366 }, { x: 798, y: 1 }],
        [{ x: 1, y: 1 }, { x: 41, y: 390 }, { x: 773, y: 390 }, { x: 798, y: 1 }]
      ].map((points, index) => ({ points, index }))
    );
    expect(polygonTransforms).toMatchSnapshot();
    const polygonTransforms2 = greedyPack(
      800,
      742,
      [
        [{ x: 1, y: 1 }, { x: 1, y: 366 }, { x: 798, y: 366 }, { x: 798, y: 1 }],
        [{ x: 1, y: 1 }, { x: 41, y: 390 }, { x: 773, y: 390 }, { x: 798, y: 1 }]
      ].map((points, index) => ({ points, index })),
      { rotationMode: RotationMode.Advanced }
    );
    expect(polygonTransforms2).toMatchSnapshot();
    const polygonTransforms3 = greedyPack(
      800,
      742,
      [
        [{ x: 1, y: 1 }, { x: 1, y: 366 }, { x: 798, y: 366 }, { x: 798, y: 1 }],
        [{ x: 1, y: 1 }, { x: 41, y: 390 }, { x: 773, y: 390 }, { x: 798, y: 1 }]
      ].map((points, index) => ({ points, index })),
      { polygonHitboxScale: 0.8 }
    );
    expect(polygonTransforms3).toMatchSnapshot();
    expect(
      greedyPack(
        742,
        800,
        [
          [{ x: 1, y: 1 }, { x: 1, y: 366 }, { x: 798, y: 366 }, { x: 798, y: 1 }],
          [{ x: 1, y: 1 }, { x: 41, y: 390 }, { x: 773, y: 390 }, { x: 798, y: 1 }]
        ].map((points, index) => ({ points, index }))
      )
    ).toMatchSnapshot();
  });
  it("normalizePolygons can be disabled", () => {
    expect(
      greedyPack(
        800,
        742,
        [
          [{ x: 1, y: 1 }, { x: 1, y: 366 }, { x: 798, y: 366 }, { x: 798, y: 1 }],
          [{ x: 1, y: 1 }, { x: 41, y: 390 }, { x: 773, y: 390 }, { x: 798, y: 1 }]
        ].map((points, index) => ({ points, index })),
        { normalizePolygons: false }
      )
    ).toMatchSnapshot();
  });
});
