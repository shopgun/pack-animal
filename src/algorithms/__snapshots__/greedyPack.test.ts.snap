// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`greedyPack gives nothing for nothingg 1`] = `Array []`;

exports[`greedyPack normalizePolygons can be disabled 1`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 94.64375%;
        height: 46.73180592991914%;
        z-index: 827;
        transform: translateX(5.28296902859407%) translateY(113.41312184571018%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.95,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.95,
      \\"e\\":40,
      \\"f\\":393.26
    }",
    "points": Array [
      Object {
        "x": 40.95,
        "y": 394.21,
      },
      Object {
        "x": 40.95,
        "y": 740.96,
      },
      Object {
        "x": 798.0999999999999,
        "y": 740.96,
      },
      Object {
        "x": 798.0999999999999,
        "y": 394.21,
      },
    ],
    "rotate": -0,
    "scale": 0.95,
    "svgTransform": "matrix(0.95,0,0,0.95,40,393.26)",
    "translateX": 40,
    "translateY": 393.26,
    "zIndex": 827,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 94.64375%;
        height: 49.80458221024258%;
        z-index: 85;
        transform: translateX(5.28296902859407%) translateY(6.023542145853065%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.95,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.95,
      \\"e\\":40,
      \\"f\\":22.259999999999998
    }",
    "points": Array [
      Object {
        "x": 40.95,
        "y": 23.209999999999997,
      },
      Object {
        "x": 78.94999999999999,
        "y": 392.76,
      },
      Object {
        "x": 774.3499999999999,
        "y": 392.76,
      },
      Object {
        "x": 798.0999999999999,
        "y": 23.209999999999997,
      },
    ],
    "rotate": -0,
    "scale": 0.95,
    "svgTransform": "matrix(0.95,0,0,0.95,40,22.259999999999998)",
    "translateX": 40,
    "translateY": 22.259999999999998,
    "zIndex": 85,
  },
]
`;

exports[`greedyPack works 1`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 97.67575877568494%;
        height: 48.22890685505281%;
        z-index: 773;
        transform: translateX(2.047590953035804%) translateY(105.74571004330544%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.980434216067101,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.980434216067101,
      \\"e\\":16,
      \\"f\\":378.42
    }",
    "points": Array [
      Object {
        "x": 16.980434216067103,
        "y": 379.40043421606714,
      },
      Object {
        "x": 16.980434216067103,
        "y": 737.258923080559,
      },
      Object {
        "x": 798.3865044215466,
        "y": 737.258923080559,
      },
      Object {
        "x": 798.3865044215466,
        "y": 379.40043421606714,
      },
    ],
    "rotate": -0,
    "scale": 0.980434216067101,
    "svgTransform": "matrix(0.980434216067101,0,0,0.980434216067101,16,378.42)",
    "translateX": 16,
    "translateY": 378.42,
    "zIndex": 773,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 96.60479773621509%;
        height: 50.83654853867456%;
        z-index: 24;
        transform: translateX(3.105435827516218%) translateY(0%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9696842934626357,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9696842934626357,
      \\"e\\":24,
      \\"f\\":0
    }",
    "points": Array [
      Object {
        "x": 24.969684293462635,
        "y": 0.9696842934626357,
      },
      Object {
        "x": 63.75705603196806,
        "y": 378.17687445042793,
      },
      Object {
        "x": 773.5659588466174,
        "y": 378.17687445042793,
      },
      Object {
        "x": 797.8080661831833,
        "y": 0.9696842934626357,
      },
    ],
    "rotate": -0,
    "scale": 0.9696842934626357,
    "svgTransform": "matrix(0.9696842934626357,0,0,0.9696842934626357,24,0)",
    "translateX": 24,
    "translateY": 0,
    "zIndex": 24,
  },
]
`;

exports[`greedyPack works 2`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 62.61266588184932%;
        height: 30.91596593272616%;
        z-index: 808;
        transform: translateX(158.78875879140813%) translateY(2.8051870536460433%) rotate(90deg)
      ",
    "matrix": "{
      \\"a\\":3.848351360516409e-17,
      \\"b\\":0.6284834718378852,
      \\"c\\":-0.6284834718378852,
      \\"d\\":3.848351360516409e-17,
      \\"e\\":795.375,
      \\"f\\":6.435000000000059
    }",
    "points": Array [
      Object {
        "x": 794.7465165281621,
        "y": 7.0634834718379444,
      },
      Object {
        "x": 565.3500493073341,
        "y": 7.063483471837959,
      },
      Object {
        "x": 565.3500493073341,
        "y": 507.96481052663245,
      },
      Object {
        "x": 794.7465165281621,
        "y": 507.96481052663245,
      },
    ],
    "rotate": 90,
    "scale": 0.6284834718378852,
    "svgTransform": "matrix(3.848351360516409e-17,0.6284834718378852,-0.6284834718378852,3.848351360516409e-17,795.375,6.435000000000059)",
    "translateX": 795.375,
    "translateY": 6.435000000000059,
    "zIndex": 808,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 61.92615239500967%;
        height: 32.587531114534976%;
        z-index: 470;
        transform: translateX(52.56706423343069%) translateY(43.28775678331294%) rotate(52.00000000000001deg)
      ",
    "matrix": "{
      \\"a\\":0.3826905530213621,
      \\"b\\":0.4898215710590342,
      \\"c\\":-0.4898215710590342,
      \\"d\\":0.3826905530213621,
      \\"e\\":260.42208245421534,
      \\"f\\":104.66957118225378
    }",
    "points": Array [
      Object {
        "x": 260.3149514361777,
        "y": 105.54208330633418,
      },
      Object {
        "x": 85.08198241506784,
        "y": 274.0015712740054,
      },
      Object {
        "x": 365.21146722670494,
        "y": 632.5509612892184,
      },
      Object {
        "x": 565.3193221942033,
        "y": 495.9298754403844,
      },
    ],
    "rotate": 52.00000000000001,
    "scale": 0.6215924958093819,
    "svgTransform": "matrix(0.3826905530213621,0.4898215710590342,-0.4898215710590342,0.3826905530213621,260.42208245421534,104.66957118225378)",
    "translateX": 260.42208245421534,
    "translateY": 104.66957118225378,
    "zIndex": 470,
  },
]
`;

exports[`greedyPack works 3`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 97.67575877568494%;
        height: 48.22890685505281%;
        z-index: 773;
        transform: translateX(2.047590953035804%) translateY(105.74571004330544%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.980434216067101,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.980434216067101,
      \\"e\\":16,
      \\"f\\":378.42
    }",
    "points": Array [
      Object {
        "x": 16.980434216067103,
        "y": 379.40043421606714,
      },
      Object {
        "x": 16.980434216067103,
        "y": 737.258923080559,
      },
      Object {
        "x": 798.3865044215466,
        "y": 737.258923080559,
      },
      Object {
        "x": 798.3865044215466,
        "y": 379.40043421606714,
      },
    ],
    "rotate": -0,
    "scale": 0.980434216067101,
    "svgTransform": "matrix(0.980434216067101,0,0,0.980434216067101,16,378.42)",
    "translateX": 16,
    "translateY": 378.42,
    "zIndex": 773,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 96.60479773621509%;
        height: 50.83654853867456%;
        z-index: 172;
        transform: translateX(3.105435827516218%) translateY(19.67088696509829%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9696842934626357,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9696842934626357,
      \\"e\\":24,
      \\"f\\":74.2
    }",
    "points": Array [
      Object {
        "x": 24.969684293462635,
        "y": 75.16968429346264,
      },
      Object {
        "x": 63.75705603196806,
        "y": 452.3768744504279,
      },
      Object {
        "x": 773.5659588466174,
        "y": 452.3768744504279,
      },
      Object {
        "x": 797.8080661831833,
        "y": 75.16968429346264,
      },
    ],
    "rotate": -0,
    "scale": 0.9696842934626357,
    "svgTransform": "matrix(0.9696842934626357,0,0,0.9696842934626357,24,74.2)",
    "translateX": 24,
    "translateY": 74.2,
    "zIndex": 172,
  },
]
`;

exports[`greedyPack works 4`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 97.2099610456744%;
        height: 41.29136409974906%;
        z-index: 943;
        transform: translateX(2.0574023263524337%) translateY(140.4652068647751%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9050161994465548,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9050161994465548,
      \\"e\\":14.84,
      \\"f\\":464
    }",
    "points": Array [
      Object {
        "x": 15.745016199446555,
        "y": 464.90501619944655,
      },
      Object {
        "x": 15.745016199446555,
        "y": 795.235928997439,
      },
      Object {
        "x": 737.0429271583507,
        "y": 795.235928997439,
      },
      Object {
        "x": 737.0429271583507,
        "y": 464.90501619944655,
      },
    ],
    "rotate": -0,
    "scale": 0.9050161994465548,
    "svgTransform": "matrix(0.9050161994465548,0,0,0.9050161994465548,14.84,464)",
    "translateX": 14.84,
    "translateY": 464,
    "zIndex": 943,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 96.1441072224409%;
        height: 43.523906556572925%;
        z-index: 246;
        transform: translateX(3.1203160408563995%) translateY(32.16623025739342%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.89509319396551,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.89509319396551,
      \\"e\\":22.259999999999998,
      \\"f\\":112
    }",
    "points": Array [
      Object {
        "x": 23.15509319396551,
        "y": 112.89509319396551,
      },
      Object {
        "x": 58.95882095258591,
        "y": 461.0863456465489,
      },
      Object {
        "x": 714.1670389353392,
        "y": 461.0863456465489,
      },
      Object {
        "x": 736.544368784477,
        "y": 112.89509319396551,
      },
    ],
    "rotate": -0,
    "scale": 0.89509319396551,
    "svgTransform": "matrix(0.89509319396551,0,0,0.89509319396551,22.259999999999998,112)",
    "translateX": 22.259999999999998,
    "translateY": 112,
    "zIndex": 246,
  },
]
`;
